pkg/util/limiter/rate_limiter.go

```diff
@@ -42,9 +50,29 @@ func NewRateLimiter(strategy RateLimiterStrategy, recheckPeriod time.Duration) *
 	}
 }
 
-// AllowN reports whether n tokens may be consumed happen at time now.
-func (l *RateLimiter) AllowN(now time.Time, tenantID string, n int) bool {
-	return l.getTenantLimiter(now, tenantID).AllowN(now, n)
+// AllowN reports whether n tokens may be consumed happen at time now. The
+// reservation of tokens can be canceled using CancelAt on the returned object.
+func (l *RateLimiter) AllowN(now time.Time, tenantID string, n int) *Reservation {
```

Comment:
This is used in the write path, which is an hot path. I think we could avoid extra allocations for `type Reservation struct` doing the following:

1. Change `type Reservation struct` to `type Reservation interface { CancelAt(now time.Time) }` (`*rate.Reservation` will already implements this interface)
2. Change this `AllowN()` return from `*Reservation` to `(bool, Reservation)` (no pointer because now it's an interface) and directly return `(ok, *rate.Reservation)`

The `type Reservation interface` wouldn't be strictly required, but it's a nice to have to abstract away the underlying `*rate.Reservation` so we don't leak it.

Reply:
Agreed - good point.